server/server.go

```diff
@@ -42,72 +55,362 @@ func HandleWebsocketConnection(serverHTTPResponse http.ResponseWriter, clientHTT
+	messageToClient.Request.Auth = &backendProtobuf.BackendMessage_Auth{}
+	// Create a byte slice to store our random bytes
+	backendRandomBytes := make([]byte, 4)
+	// Read random bytes into our slice
+	_, randomBytesErr := rand.Read(backendRandomBytes)
+	// If there is an error reading random bytes
+	if randomBytesErr != nil {
+		return "", randomBytesErr
+	} // if randomBytesErr != nil
+	// Set the byte sequence that the client needs to sign
+	messageToClient.Request.Auth.ToSign = backendRandomBytes
 	// Prepare the data to be sent to the client
-	protobufAuthBytes, protobufAuthBytesErr := proto.Marshal(&protobufAuth)
-	if protobufAuthBytesErr != nil {
+	messageToClientBytes, messageToClientBytesErr := proto.Marshal(&messageToClient)
+	if messageToClientBytesErr != nil {
 		// If there is an error while trying to perform protobuf marshaling, terminate the connection
-		return websocket.BinaryMessage, protobufAuthBytesErr
-	}
+		return "", messageToClientBytesErr
+	} // if messageToClientBytesErr != nil
 	// Send the protobuf data to the client containing the sequence of bytes he needs to sign
-	websocketConnection.WriteMessage(websocket.BinaryMessage, protobufAuthBytes)
+	websocketConnection.WriteMessage(websocket.BinaryMessage, messageToClientBytes)
 	// Read the response from the client which should contain his IdenetityPublicKey and the signed byte sequence
-	_, clientMessage, readMessageErr := websocketConnection.ReadMessage()
+	_, messageFromClientProto, readMessageErr := websocketConnection.ReadMessage()
 	if readMessageErr != nil {
 		// If there is an error while reading the response from the client, terminate the connection
-		return websocket.BinaryMessage, readMessageErr
+		return "", readMessageErr
 	}
 	// Unmarshal the response from the client into our protobuf Auth structure
-	protoUnmarshalErr := proto.Unmarshal(clientMessage, &protobufAuth)
+	protoUnmarshalErr := proto.Unmarshal(messageFromClientProto, &messageFromClient)
 	if protoUnmarshalErr != nil {
 		// If there is an error while unmarshaling the response from the client, terminate the connection
-		return websocket.BinaryMessage, protoUnmarshalErr
+		return "", protoUnmarshalErr
 	}
 	// Create a [32]byte{} identityPublicKey to satisfy ed25519.Verify() type requirements
 	identityPublicKey := [32]byte{}
 	// Create a [64]byte{} signedByteSequenceFromClient to satisfy ed25519.Verify() type requirements
 	signedByteSequenceFromClient := [64]byte{}
+	// Create a string representation of the Identity Public Key
+	identityPublicKeyString := string(messageFromClient.Response.Auth.IdentityPublicKey)
 	// Decode the hex representation of the identityPublicKey here as transmitting it over the network without hex encoding can cause issues with some clients
-	identityPublicKeyDecoded, identityPublicKeyDecodedErr := hex.DecodeString(string(protobufAuth.GetIdentityPublicKey()))
+	identityPublicKeyDecoded, identityPublicKeyDecodedErr := hex.DecodeString(identityPublicKeyString)
```

Comment:
I don't think `hex.DecodeString` is gonna work since `identityPublicKeyString` is not a hex string as far as I remember (since you used `string()`). 

```go
import (
	"math/rand"
	"fmt"
)

func main()  {
	
	b := make([]byte, 32)
	rand.Read(b)
	fmt.Println(string(b))
	
}
```

does e.g. output `R��!�eO?_�br�f�M|M{����I`

Reply:
I mentioned this on a similar comment before. 

In the test client I wrote I already pre-hex encoded the identityPublicKey before sending it over the protobuf. 

The reason I did this was because I remember there was some error in the past caused exactly by the strange characters being output that you mentioned. 

Is there any downside of the client himself already sending a hex encoded representation?

Reply:
it's not really a problem. I have the feeling go just prefere to work with byte slices instead of strings. It's also better to convert from bytes to hex then converting from hex to bytes since `hex.EncodeToString` just returns a string. But `hex.DecodeString` returns an byte slice and an error.